src/main/java/org/wildfly/security/auth/provider/jdbc/JdbcSecurityRealm.java

```diff
@@ -0,0 +1,254 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.wildfly.security.auth.provider.jdbc;
+
+import org.wildfly.security._private.ElytronMessages;
+import org.wildfly.security.auth.principal.NamePrincipal;
+import org.wildfly.security.auth.provider.jdbc.mapper.PasswordKeyMapper;
+import org.wildfly.security.auth.spi.AuthenticatedRealmIdentity;
+import org.wildfly.security.auth.spi.CredentialSupport;
+import org.wildfly.security.auth.spi.RealmIdentity;
+import org.wildfly.security.auth.spi.RealmUnavailableException;
+import org.wildfly.security.auth.spi.SecurityRealm;
+import org.wildfly.security.password.Password;
+import org.wildfly.security.password.PasswordFactory;
+import org.wildfly.security.password.interfaces.ClearPassword;
+
+import javax.sql.DataSource;
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+import java.security.Principal;
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.util.List;
+
+/**
+ * @author <a href="mailto:psilva@redhat.com">Pedro Igor</a>
+ */
+public class JdbcSecurityRealm implements SecurityRealm {
+
+    private final List<QueryConfiguration> queryConfiguration;
+
+    public static JdbcSecurityRealmBuilder builder() {
+        return new JdbcSecurityRealmBuilder();
+    }
+
+    public JdbcSecurityRealm(List<QueryConfiguration> queryConfiguration) {
+        this.queryConfiguration = queryConfiguration;
+    }
+
+    @Override
+    public RealmIdentity createRealmIdentity(Principal principal) throws RealmUnavailableException {
+        if (principal instanceof NamePrincipal == false) {
+            throw ElytronMessages.log.invalidPrincipalType(NamePrincipal.class, principal == null ? null : principal.getClass());
+        }
+
+        return new JdbcRealmIdentity(principal);
+    }
+
+    @Override
+    public CredentialSupport getCredentialSupport(Class<?> credentialType) throws RealmUnavailableException {
+        for (QueryConfiguration configuration : this.queryConfiguration) {
+            for (ColumnMapper mapper : configuration.getColumnMappers()) {
+                if (KeyMapper.class.isInstance(mapper)) {
+                    KeyMapper credentialMapper = (KeyMapper) mapper;
+
+                    if (credentialType.equals(credentialMapper.getCredentialType())) {
+                        return CredentialSupport.FULLY_SUPPORTED;
```

Comment:
We should probably still look at where some accounts have different credential types associated with them.